adala/runtimes/_litellm.py

```diff
@@ -167,25 +245,36 @@ async def batch_to_batch(
             lambda row: input_template.format(**row, **extra_fields), axis=1
         ).tolist()
 
-        responses: List[Union[ConstrainedLLMResponse, ErrorLLMResponse]] = (
-            await parallel_async_get_llm_response(
-                user_prompts=user_prompts,
-                system_prompt=instructions_template,
-                instruction_first=instructions_first,
-                response_model=response_model,
-                inference_settings=LiteLLMInferenceSettings(
-                    **self.dict(include=LiteLLMInferenceSettings.model_fields.keys())
-                ),
+        tasks = [
+            asyncio.ensure_future(
+                async_instructor_client.chat.completions.create(
+                    messages=get_messages(
+                        user_prompt, self.system_prompt, self.instruction_first
+                    ),
+                    response_model=response_model,
+                    **self.as_inference_settings().dict(),
+                )
             )
-        )
+            for user_prompt in user_prompts
+        ]
+        responses = await asyncio.gather(*tasks, return_exceptions=True)
 
         # convert list of LLMResponse objects to the dataframe records
         df_data = []
         for response in responses:
-            if isinstance(response, ErrorLLMResponse):
+            if isinstance(response, Exception):
+                error_message = type(response).__name__
+                # error_details = str(response)
+                error_details = traceback.format_exc()
                 if self.verbose:
-                    print_error(response.adala_message, response.adala_details)
-                df_data.append(response.model_dump(by_alias=True))
+                    print_error(error_message, error_details)
+                # TODO change this format
+                error_dct = {
```

Comment:
The response classes were used as the output types for `*_get_llm_response`, which were specific to litellm, and one of 3 things happened to them immediately after being returned:
- `runtime.get_llm_response`: raised as an error or converted to string
- `runtime.record_to_record`: coerced to dict
- `runtime.batch_to_batch`: coerced to dataframe row

For UnconstrainedResponse (just a string, and not used outside of a legacy call in learn() anyway), and ConstrainedResponse (is already a pydantic response_model turned into a dict field, which is worse than using the underlying response_model), I think this makes a lot of sense. I'm more on the fence about the ErrorResponse in particular:
> Bring back LLMErrorResponse model? it's not useful now because it'd get immediately coerced into a dataframe row or a dict anyway, but hoping to build towards an alternate inference path from the server which doesn't convert to and from dataframes, so we can use pydantic models defined in the rest of the adala lib directly in ResultHandlers.

I removed it here, but I can see the case for keeping it, especially as we evolve our error handling in the future and gain the ability to inspect the underlying LiteLLM error class and handle each one differently. This is also the one that I think is most likely to have applications for other runtimes outside LiteLLM (specifically sglang, as Nikolai brought up).

Reply:
Sure, I totally see the motivation here, I just want to avoid us always having to turn small-ish changes into minor refactors (esp ones that negate parts of the last minor refactor).
It's great that we're incrementally improving the code here, but it's starting to feel like we're thrashing a bit with the design.
As for this PR, I'm good with this (and agree that it feels simpler), mainly just want us to be wary about continuing to go back and forth.